architecture-decision-records/0027-step-functions-for-backend-checks.md

```diff
@@ -0,0 +1,42 @@
+# 26. Step functions for backend checks
+
+**Date:** 2023-01-18
+
+## Context
+Once the files in TDR are uploaded, we run antivirus, a server side checksum and a Droid file format check.
+
+We also need to set various file and consignment statuses based on the results of the checks.
+
+We also need to now the status of any redacted files in the consignment.
+
+## Options considered
+### Use the existing process
+The existing process is driven by an SNS notification from S3 which triggers the backend checks lambdas. 
+Communication between lambdas is handled by SQS queues.
+
+#### Advantages
+* It is event driven, so we don't have to manually trigger the process.
+* It is cheaper, as there is a cost to the step functions.
+
+#### Disadvantages
+* It is controlled by a queue. If a large consignment is uploaded, followed by a small judgment consignment, the judgment consignment will need to wait for the large consignment to finish and this may take several hours.
+* It is difficult to check statuses which apply across all files in a consignment, for example, all files need to be checked for redacted status.
+* The API is called once for each file. This causes a lot of load on the database. 
+
+### Use a step function
+The step function will run once per consignment. 
+It will be triggered by API Gateway and authenticated with the existing consignment export authoriser lambda.
+
+#### Advantages
+* It is run once per consignment which won't block smaller transfers with larger ones. 
+* We can run a lambda once all the file checks are complete which means updating consignment statuses which apply to all files easier.
```

Comment:
Maybe reword slightly: `We can run a lambda once all the file checks are complete, this makes it easier to update consignment statuses, which require to all files statuses.`

Reply:
Did you mean `require to all file statuses` or should that be `require all file statuses`? I've changed it to the second one for now.